packages/tpetra/core/src/Tpetra_Details_WrappedDualView.cpp

```diff
@@ -0,0 +1,60 @@
+// @HEADER
+// ***********************************************************************
+//
+//          Tpetra: Templated Linear Algebra Services Package
+//                 Copyright (2008) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// ************************************************************************
+// @HEADER
+
+namespace Tpetra {
+namespace Details {
+
+bool wdvTrackingEnabled = true;
+
+void enableWDVTracking()
+{
+  if(wdvTrackingEnabled)
+    throw std::runtime_error("WrappedDualView refcount tracking is already enabled!");
+  wdvTrackingEnabled = true;
+}
+
+void disableWDVTracking()
+{
+  if(!wdvTrackingEnabled)
```

Comment:
I don't know how `Tpetra` positions w.r.t. task-based computing, but I guess such tricks based on global variable are dangerous from that view point?

Reply:
It seems you are mimicking the `Kokkos::View` ref counting behavior. See for instance `packages/kokkos/core/src/Kokkos_Parallel.hpp`:

```c++
Kokkos::Impl::shared_allocation_tracking_disable();
Impl::ParallelFor<FunctorType, ExecPolicy> closure(functor, inner_policy);
Kokkos::Impl::shared_allocation_tracking_enable();
```
so when the view is copied on the device, it stops reference counting.

Note that in this case, the variable is **thread_local**:
```c++
static inline thread_local int t_tracking_enabled = 1;
```

Reply:
I think I would be fine with the "global flag" approach as long as it is `thread_local`. The use case I see is:
- Thread 1 works on a problem with `Tpetra` object `OA`
- Thread 2 works on another problem with `Tpetra` object `OB`

These threads do uncorrelated things so the global flag should be *per-thread*.

I agree that having the flag *per-WDV* could be cumbersome in some cases (e.g. if several WDV are used in a parallel region).

As a side question, do you think using `execution_space::in_parallel` could be an alternative? I don't know what's the status of this `in_parallel` because it seems nobody uses that.
```

Reply:
I also wanted to say, to concurrently use Tpetra objects, the one approach that would work is to split the MPI comm. Say the global comm has 8 ranks, over 4 nodes. If you make sub-communicators for the even and odd ranks, you could then put ``OA`` on the evens and ``OB`` on the odds, and effectively get similar behavior without thread safety problems.